pkg/mutation/assignmetadata.go

```diff
@@ -0,0 +1,72 @@
+package mutation
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// MetadataMutator is a mutator wrapping the AssignMetadata type.
+type AssignMetadataMutator struct {
+	//assignMetadata *mutationsv1.AssignMetadata
+	Path  parser.Path
+	Value string
+}
+
+func (m AssignMetadataMutator) Matches(obj metav1.Object, ns *corev1.Namespace) bool {
+	// TODO: add match logic
+	return false
+}
+
+func (m AssignMetadataMutator) Mutate(obj *unstructured.Unstructured) error {
+	return mutate(obj.Object, m.Path.Nodes, m.Value)
+}
+
+func mutate(current interface{}, remainingPaths []parser.Node, value string) error {
+	pathEntry := remainingPaths[0]
+	if len(remainingPaths) == 1 {
+		// TODO: handle types other than object
+		current.(map[string]interface{})[pathEntry.(parser.Object).Reference] = value
+		return nil
+	}
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		next, ok := current.(map[string]interface{})[pathEntry.(parser.Object).Reference]
+		if !ok {
+			// TODO: Add pathTests and add path if required.
+			// next := NewEmptyOfCorrectType(remainingPaths[1])
+			return fmt.Errorf("Path specified by PathEntry does not exist for resource")
+		}
+		if err := mutate(next, remainingPaths[1:], value); err != nil {
+			return err
+		}
+		return nil
+	case parser.ListNode:
+		for _, listElement := range current.([]interface{}) {
+			if pathEntry.(parser.List).Glob {
+				if err := mutate(listElement, remainingPaths[1:], value); err != nil {
+					return err
+				}
+			} else if elementValue, ok := listElement.(map[string]interface{})[pathEntry.(parser.List).KeyField]; ok {
```

Comment:
I'm starting to think we can use the same core `mutate()` logic for both assign and assign metadata, as the two objects mainly differ in terms of how they are validated. Otherwise, we don't need to handle lists for metadata yet... both annotations and labels are "objects" (`map[string]string`).

If we are writing a generalized mutation function here, we should be sure it also handles the case:

`spec.containers[name: "foo"] = <<contents of a container object>>`

In such a case, the object may not be found in the for-loop, thus it will need to be created and added to the list.

Reply:
> I'm starting to think we can use the same core mutate() logic for both assign and assign metadata, as the two objects mainly differ in terms of how they are validated. Otherwise, we don't need to handle lists for metadata yet... both annotations and labels are "objects" (map[string]string).

Renamed to BaseMutator to cover both types.

> If we are writing a generalized mutation function here, we should be sure it also handles the case:
> spec.containers[name: "foo"] = <<contents of a container object>>
> In such a case, the object may not be found in the for-loop, thus it will need to be created and added to the list.

Modified the function to add missing path. Note that since the value can only be a simple string value, we will need multiple Assigns to create a whole object.
Note that path tests will only be added in a follow up PR